### PR TITLE
Driver `/solve` API

### PR DESCRIPTION
**Pull request by sistemd**
Progress towards https://github.com/cowprotocol/services/issues/895, the only thing missing now are tests :tada: .

Implements the auction and solution DTOs for `/solve`.

The auction DTO is very similar to the auction DTO from the driver-solver interface, with a few additional fields. The solution DTO is very straightforward.

### CODE REVIEW

File: crates/driver/src/api/solve/dto/auction.rs [hunk 1]
```diff
@@ -1,84 +1,211 @@
...
+                            },
+                            data: order.signature,
+                            signer: order.owner.into(),
+                        },
+                        reward: order.reward,
+                    })
+                })
+                .try_collect()?,
+            // TODO #899
+            liquidity: Default::default(),
```
Comment (nlordell) [inline crates/driver/src/api/solve/dto/auction.rs:123]: I don't think that the protocol will ever send liquidity as part of the `Auction` DTO - I only think this exists in the driver <> solver `Auction` DTO (as it is populated by the driver, and doesn't come from the protocol). I think we can remove it here.

File: crates/driver/src/api/solve/dto/auction.rs [hunk 2]
```diff
@@ -1,84 +1,211 @@
...
     valid_to: u32,
-    #[serde_as(as = "serialize::Hex")]
-    app_data: [u8; 32],
-    kind: OrderKind,
+    kind: Kind,
+    receiver: Option<H160>,
+    owner: H160,
     partially_fillable: bool,
-
+    // TODO Always zero if the order is not partially fillable, is that OK?
```
Comment (nlordell) [inline crates/driver/src/api/solve/dto/auction.rs:172]: 👍

File: crates/driver/src/api/solve/dto/auction.rs [hunk 3]
```diff
@@ -87,56 +214,151 @@ enum SellTokenBalance {
...
+    available_balance: U256,
+    trusted: bool,
 }
 
-#[derive(Debug, Default, Deserialize)]
-struct Interactions {
-    pre: Vec<InteractionData>,
+#[derive(Debug, Deserialize)]
+#[serde(tag = "kind", rename_all = "lowercase")]
+enum Liquidity {
```
Comment (nlordell) [inline crates/driver/src/api/solve/dto/auction.rs:256]: Shouldn't this exist in the `solver::dto` and not in `api::solve::dto`? It is my understanding that the former is the DTO between the driver and solver, and that the latter is the DTO between the protocol and the driver.

File: crates/driver/src/solver/dto/auction.rs [hunk 4]
```diff
@@ -91,6 +91,8 @@ struct Order {
     partially_fillable: bool,
     class: Class,
     reward: f64,
+    // TODO Was the intention to have the driver verify signatures, and the solvers blindly trust
+    // the orders and the driver auth? Or did we just forget to add the signature here?
```
[the remark this answers is not in the document]
Comment (nlordell) [reply crates/driver/src/solver/dto/auction.rs:95]: The solver engine doesn't need the order signature - as it is inconsequential to the settlement from a purely "mathematical price-finding" perspective.

> blindly trust the orders and the driver auth?

These signatures are only important for ensuring user orders are authorized. As such, even if we gave an order to the driver with a bad signature which was then relayed to the solver (with signature omitted) and then used in a solution - the worse that can happen is that the solution fails to simulate and/or revert.

Furthermore, as the driver is intended to be co-located with the solver, I don't think we need strong trust assumptions here, and I never imagined that solver engines would verify signatures for all the orders it receives.

File: crates/driver/src/api/solve/dto/auction.rs (hunk 1 above)
Comment (nlordell) [reply crates/driver/src/api/solve/dto/auction.rs:123]: I see it was add in #999 in the solver DTO (where I think it should be) - so I would remove this from here.

File: crates/driver/src/api/solve/dto/auction.rs (hunk 3 above)
Comment (nlordell) [reply crates/driver/src/api/solve/dto/auction.rs:256]: Same comment as ☝️, so closing this one for less noise.

**Review by vkgnosis — APPROVED** [review]

File: crates/driver/src/api/solve/dto/auction.rs (hunk 1 above)
Comment (nlordell) [reply crates/driver/src/api/solve/dto/auction.rs:123]: I understood that this was the DTO that was being sent from the `autopilot` to the `driver` which doesn't have liquidity 🤦.

File: crates/driver/src/api/solve/dto/auction.rs (hunk 3 above)
Comment (nlordell) [inline crates/driver/src/api/solve/dto/auction.rs:256]: This shouldn't be here though right?

There is no liquidity DTO from `autopilot` -> `driver`.

**Review by nlordell — APPROVED** [review]
LGTM. I don't think we need `Liquidity` DTO from `autopilot` -> `driver`.

**Pull request merged**